src/osd/OSDMap.cc

```diff
@@ -5196,6 +5273,229 @@ int OSDMap::calc_desired_primary_distribution(
   return 0;
 }
 
+//
+// For read balancing with different osd sizes - calculate the desired number of primaries
+// per OSD for a given pool with constraints (forced*) that are set by previous
+// assignments.
+//
+float OSDMap::calc_desired_prims_for_osdsizeopt(int npgs, int forced_primaries,
+                                                int forced_secondaries, int iops_per_osd,
+                                                int write_ratio) const {
+  int forced_iops_per_pg = forced_primaries * 100 + forced_secondaries * write_ratio;
+  int pgs_left = npgs - forced_primaries - forced_secondaries;
+  int iops_left = iops_per_osd - forced_iops_per_pg;
+  if (pgs_left <= 0)
+    return float(forced_primaries);
+  else
+    return float(forced_primaries) + float(iops_left - pgs_left * write_ratio) / float(100 - write_ratio);
+}
+
+int OSDMap::calc_desired_primary_distribution_osdsize_opt(
+  CephContext *cct,
+  int64_t pid,
+  const vector<uint64_t> &osds,
+  map<uint64_t, float>& desired_primary_distribution) const
+{
+  const pg_pool_t* pool = get_pg_pool(pid);
+  if (!pool->is_replicated()) {   // read balancing works only for replicated pools
+    ldout(cct, 10) << __func__ <<" skipping erasure pool "
+                   << get_pool_name(pid) << dendl;
+    return -EINVAL;
+  } else if (pool->get_size() <= 1) {
+    ldout(cct, 10) << __func__ << " skipping replicated pool "
+                   << get_pool_name(pid) <<  " with a single replica" << dendl;
+    return -EINVAL;
+  } else {
+    int replica_count = pool->get_size();
+    int pg_num = pool->get_pg_num();
+    map<uint64_t,set<pg_t>> pgs_by_osd;
+    pgs_by_osd = get_pgs_by_osd(cct, pid);
+    int sum_pgs = 0;
+    for (auto& [_, pgs] : pgs_by_osd) {
+      sum_pgs += pgs.size();
+    }
+    if (sum_pgs != pg_num * replica_count) {
+      ldout(cct, 10) << __func__ << " Some of the PGs for pool '"
+                     << get_pool_name(pid) << "' don't have" << replica_count << " replicas "
+                     << "- can't perform osd-size-optimized read balancing " << dendl;
+      return -EINVAL;
+    }
+    ldout(cct, 20) << __func__ << " calculating OSD size optimized primary distribution for replicated pool "
+                   << get_pool_name(pid) << dendl;
+
+    int64_t read_ratio = 0;
+    {
+      // new scope since it is not allowed to use def_read_ratio after std::move
+      uint64_t def_read_ratio = cct->_conf.get_val<uint64_t>("osd_pool_default_read_ratio");
+      read_ratio = pool->opts.value_or<int64_t>(pool_opts_t::key_t::READ_RATIO, std::move(def_read_ratio));
+    }
+    int write_ratio = 100 - read_ratio;
+    ldout(cct, 30) << __func__ << " Pool: " << pid << " read ratio: " << read_ratio << " write ratio: " << write_ratio << dendl;
+    int num_osds = osds.size();
+    if (pg_num != (pool->get_pgp_num_mask() + 1)) {
```

Comment:
Seeing this compile warning. One of them should be cast as (int) or (uint):
```
/home/lflores/ceph/src/osd/OSDMap.cc:5335:16: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
 5335 |     if (pg_num != (pool->get_pgp_num_mask() + 1)) {
      |         ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
Changed declaration to `auto`